README.md
# Calibre's "Find Duplicates" Plugin

There is a plugin for Calibre that allows for many methods of finding duplicate books.
The thread where discussion and development of this plugin occurs appears
to be [here](https://www.mobileread.com/forums/showthread.php?t=131017). Thus, I
believe the primary developer to be "kiwidude" on the mobileread.com forums.

However, I was unable to find a public git (or other version control) repository
for this plugin, and I wanted to contribute further functionality. So, until
I can figure out how the primary developer prefers to receive contributions (if at all),
I've uploaded the entire plugin here, along with a commit for my change.


changelog.txt
[B]Version 1.8.5[/B] - 12 February 2021 - by capink
Fix: bug when exporting duplicates to json file in Windows.
Fix: bug with calibre 2.x failing to import missing class. https://www.mobileread.com/forums/showpost.php?p=4090981&postcount=820

[B]Version 1.8.4[/B] - 7 January 2021 - by capink
Fix: update the plugin to use calibre.library.db for target database instead of the deprecated LibraryDatabase2. Also update to use db.new_api.get_proxy_metadata to improve performance when using templates in advanced mode.

[B]Version 1.8.3[/B] - 21 October 2020 - by capink
Update: When changing libraries in library compare, restore the last used match rules from previous library if possible (all columns in match rules present in the newly selected library).
Fix: Remove invalid locations from saved location list in library compare dialog before restoring.

[B]Version 1.8.2[/B] - 17 October 2020 - by capink
Update: Restore last used match rules (and sort filters).
Fix: Allow dialog size to be reduced.

[B]Version 1.8.1[/B] - 15 October 2020 - by capink
Minor fixes.

[B]Version 1.8.0[/B] - 11 October 2020 - by capink
Update: Add advanced mode. It allows the user to match books without restrictions on the type nor the number of columns used. It also allows for user defined algorithms by using templates. It comes with a sort dialog allowing you to sort books based on columns and templates. To complement the sort feature, it adds extra marks to first and last books in each duplicate group: "first_duplicate", "last_duplicate".
Update: Mark records with deleted formats in binary search as "deleted_binary_duplicate"
Update: Option to export duplicate groups to json file. For advanced mode, the sorting of books is retained in the json file.
Update: Update Spanish translation. Thanks to @dunhill.
Fix: Mark exemptions only when showing them and remove the marks afterwards.
Fix: Restore state if the user exits calibre with the duplicates restriction still on. Thanks to @chaley
Fix: Remember last sort in library view and revert back to it.
Code refactoring.
Calibre minimum version bumped to 2.0.0
Schema version bumped to 1.6

[B]Version 1.7.2[/B] - 25 June 2020 - by davidfor
Fix: A couple of errors with translations.

[B]Version 1.7.0[/B] - 21 June 2020 - by davidfor
New: Make translatable.
New: Add Spanish translation. Thanks to @dunhill.
Change: Use delete key to remove entry from library list in cross library search options.
Update: Changes for Python 3 support in calibre.

[B]Version 1.6.3[/B] - 12 Jun 2017
fixed compatibility with Calibre 2.99b11+

[B]Version 1.6.1[/B] - 03 Jan 2013
Fix for when comparing library duplicates to ensure saved searches are not corrupted.

[B]Version 1.6.0[/B] - 29 Oct 2012
Change "ISBN Compare" to "Identifier" with a dropdown allowing comparison of any identifier field.
Add a context menu to the metadata variations list to allow choosing the selected name on the right side.

[B]Version 1.5.3[/B] - 14 Aug 2012
When using "Find library duplicates" display all duplicate matches for the current library as marked:duplicate (except for author duplicates)

[B]Version 1.5.2[/B] - 21 Jul 2012
When using "Find library duplicates" clear the current search in order to compare the entire restricted library
When using "Find metadata variations" and showing books, fire the search again to ensure results reflect the search

[B]Version 1.5.1[/B] - 21 Jul 2012
Add a "Save log" button for the "Find library duplicates" result screen.

[B]Version 1.5.0[/B] - 20 Jul 2012
Add a "Find library duplicates" option for cross-library duplicate comparisons into a log report
If currently running a duplicate book search and execute a metadata variation search, clear search first

[B]Version 1.4.0[/B] - 17 Jul 2012
Now requires calibre 0.8.59
Add a Find metadata variations option to search for author, series, publisher and tag variations, and allow renaming them from the dialog.
Fix bug of fuzzy author comparisons which will no longer compute a reverse hash to reduce the false positives it generated

[B]Version 1.3.0[/B] - 22 Jun 2012
Now requires calibre 0.8.57
Store configuration in the calibre database rather than a json file, to allow reuse from different computers (not simultaneously!)
Add a support option to the configuration dialog allowing viewing the plugin data stored in the database
Add an option to allow automatic removal of binary duplicates (does not delete books records, only the newest copies of that format).

[B]Version 1.2.3[/B] - 02 Dec 2011
Make the languages comparison optional (default false) via a checkbox on the Find Duplicates dialog

[B]Version 1.2.2[/B] - 25 Nov 2011
Take the languages field into account when doing title based duplicate comparisons

[B]Version 1.2.1[/B] - 12 Nov 2011
When selecting ISBN or Binary compare, hide the Title/Author groupbox options
Some cosmetic additions to the text for ISBN/Binary options

[B]Version 1.2.0[/B] - 11 Sep 2011
Fix bug for when switching to an ignore title search where author search was previously set to ignore.
Remove customisation of shortcuts on tab, to use Calibre's centrally managed shortcuts instead.

[B]Version 1.1.4[/B] - 04 Jul 2011
Additional fix for stuff broken by Calibre 0.8.8 in the tag view
Fix for removing an author exemption

[B]Version 1.1.3[/B] - 03 Jul 2011
Preparation for deprecation of db.format_abspath() for networked backend

[B]Version 1.1.2[/B] - 03 Jul 2011
Fix for issue with Calibre 0.8.8 tag browser search_restriction refactoring

[B]Version 1.1.1[/B] - 12 Jun 2011
Add van to list of ignored author words
Fix bug of error dialog not referenced correctly

[B]Version 1.1[/B] - 3 May 2011
Add support for binary comparison searches to find book formats with exactly the same content
Replace how exemptions are stored in the config file to make more scalable
No longer calculate exemption preview detailed messages for the confirmation dialog for performance

advanced/common.py
#!/usr/bin/env python
# ~*~ coding: utf-8 ~*~
from __future__ import (unicode_literals, division, absolute_import,
                        print_function)

__license__ = 'GPL v3'
__copyright__ = '2020, Ahmed Zaki <azaki00.dev@gmail.com>'
__docformat__ = 'restructuredtext en'

try:
    from PyQt5.Qt import (QGridLayout, QLabel, QDialogButtonBox, QPlainTextEdit, QCheckBox,
                          QWidget, QSize, QSizePolicy, QPainter)

except ImportError:
    from PyQt4.Qt import (QGridLayout, QLabel, QDialogButtonBox, QPlainTextEdit, QCheckBox,
                          QWidget, QSize, QSizePolicy, QPainter)

from functools import partial
from collections import OrderedDict, defaultdict
import re
import traceback

# python3 compatibility
from six.moves import range
from six import text_type as unicode, string_types as basestring

from calibre.gui2.dialogs.message_box import MessageBox
from calibre.ebooks.metadata.book.formatter import SafeFormat

TEMPLATE_PREFIX = 'TEMPLATE: '
TEMPLATE_ERROR = 'FD template error'

def truncate(string, length=22):
    return (string[:length] + '...') if len(string) > length else string

def to_list(string, sep=','):
    if string:
        return [a.strip().replace('|',',') for a in string.split(sep)]
    return []

STANDARD_FIELD_KEYS = [
    'title',
    'authors',
    'tags',
    'series',
    'languages',
    'publisher',
    'pubdate',
    'rating',
    'timestamp'
]

def get_cols(db):
    custom_fields = sorted([k for k,v in db.field_metadata.custom_field_metadata().items() if v['datatype'] not in ['comments']])
    all_metadata = OrderedDict()
    for field in STANDARD_FIELD_KEYS + custom_fields:
        all_metadata[field] = db.field_metadata.all_metadata()[field]
    return update_metadata(all_metadata)


def update_metadata(metadata):
    for column, meta in metadata.items():
        if column == 'publisher':
            meta['icon_name'] = 'publisher.png'
            meta['delegate'] = 'publisher'
            meta['soundex_length'] = 6
        elif meta['datatype'] == 'series':
            meta['icon_name'] = 'series.png'
            meta['delegate'] = 'series'
            meta['soundex_length'] = 6
        elif meta['is_multiple'] != {}:
            if column == 'authors' or meta['display'].get('is_names'):
                meta['icon_name'] = 'user_profile.png'
                meta['delegate'] = 'authors'
                meta['soundex_length'] = 8
            else:
                meta['icon_name'] = 'tags.png'
                meta['delegate'] = 'tags'
                meta['soundex_length'] = 4
        else:
            meta['icon_name'] = 'column.png'
            meta['delegate'] = 'title'
            meta['soundex_length'] = 6
    return metadata


class RestoreError(Exception):
    def __init__(self, error_msg):
        self.error_msg = error_msg

def do_udpate_mi(mi, field_name, value, all_metadata):
    if field_name.startswith('identifier:'):
        # 'identifier:' has no entry in all_metadata and would raise an exception
        pass
    # all composite fields (even those where ['is_multiple'] != {}) are of string type
    elif not all_metadata[field_name]['datatype'] == 'composite':
        if all_metadata[field_name]['is_multiple'] != {}:
            # for fields with multiple items, when we update a single item, we must put it in a list
            # because mi expect multiple value field, if you don't do this it will treat the string
            # value as iterable and split it into letters.
            value = [value]
    setattr(mi, field_name, value)

def algorithm_caller(algorithm, field_name, hash_, all_metadata, mi, settings, **kwargs):
    if not hasattr(algorithm, '__call__'):
        # handle the code for templates here
        
        # update the mi to presist the hash as this the only way a template can see the result of the previous algorithm/template
        do_udpate_mi(mi, field_name, hash_, all_metadata)

        template = algorithm.lstrip(TEMPLATE_PREFIX)
        hash_ = SafeFormat().safe_format(template, mi, TEMPLATE_ERROR, mi)
    else:
        kwargs2 = OrderedDict()
        kwargs2.update(settings)
        kwargs2.update(kwargs)
        hash_ = algorithm(hash_, mi, **kwargs2)
    return hash_

def get_field_value(book_id, db, field_name, mi):
    # LibraryDatabase2 does not support: field_value = db.new_api.field_for(field_name, book_id)        
    if field_name == 'title':
        field_value = db.title(book_id, index_is_id=True)
    elif field_name == 'authors':
        field_value = to_list(db.authors(book_id, index_is_id=True))
    elif field_name == 'tags':
        field_value = to_list(db.tags(book_id, index_is_id=True))
    elif field_name == 'series':
        field_value = db.series(book_id, index_is_id=True)
    elif field_name == 'publisher':
        field_value = db.publisher(book_id, index_is_id=True)
    elif field_name == 'pubdate':
        field_value = db.pubdate(book_id, index_is_id=True)
    elif field_name == 'timestamp':
        field_value = db.timestamp(book_id, index_is_id=True)
    elif field_name == 'rating':
        field_value = db.rating(book_id, index_is_id=True)
    elif field_name == 'languages':
        field_value = to_list(db.languages(book_id, index_is_id=True))
    elif field_name.startswith('#'):
        label = db.field_metadata.all_metadata()[field_name]['label']
        field_value = db.get_custom(book_id, label, index_is_id=True)
    elif field_name.startswith('identifier:'):
        identifier_type = field_name.split(':')[-1]
        identifiers = self.db.get_identifiers(book_id, index_is_id=True)
        field_value = identifiers.get(identifier_type)

    return field_value

def composite_to_list(field_name, field_value, db, composite_has_names):
    all_metadata = db.field_metadata.all_metadata()

    #{ compoiste fields with multiple items are currently returned as string, convert to list with items
    if all_metadata[field_name]['datatype'] == 'composite':
        # test first, maybe it will change in future calibre releases
        if isinstance(field_value, basestring):
            if composite_has_names:
                SEP = '&'
            else:
                SEP = all_metadata[field_name]['is_multiple']['list_to_ui']
            field_value = to_list(field_value, sep=SEP)
    #}
    return field_value

# copied from calibre.gui2.dialogs.message_box
# cannot import because it is absent in calibre 2.x
# https://www.mobileread.com/forums/showpost.php?p=4090981&postcount=820
class Icon(QWidget):

    def __init__(self, parent=None, size=None):
        QWidget.__init__(self, parent)
        self.pixmap = None
        self.setSizePolicy(QSizePolicy.Policy.Fixed, QSizePolicy.Policy.Fixed)
        self.size = size or 64

    def set_icon(self, qicon):
        self.pixmap = qicon.pixmap(self.size, self.size)
        self.update()

    def sizeHint(self):
        return QSize(self.size, self.size)

    def paintEvent(self, ev):
        if self.pixmap is not None:
            x = (self.width() - self.size) // 2
            y = (self.height() - self.size) // 2
            p = QPainter(self)
            p.drawPixmap(x, y, self.size, self.size, self.pixmap)

class MessageBox2(MessageBox):
    def setup_ui(self):
        self.setObjectName("Dialog")
        self.resize(497, 235)
        self.gridLayout = l = QGridLayout(self)
        l.setObjectName("gridLayout")
        self.icon_widget = Icon(self)
        l.addWidget(self.icon_widget)
        self.msg = la = QLabel(self)
        la.setWordWrap(True), la.setMinimumWidth(400)
        la.setOpenExternalLinks(True)
        la.setObjectName("msg")
        l.addWidget(la, 0, 1, 1, 1)
        self.det_msg = dm = QPlainTextEdit(self)
        dm.setReadOnly(True)
        dm.setObjectName("det_msg")
        l.addWidget(dm, 1, 0, 1, 2)
        self.bb = bb = QDialogButtonBox(self)
        bb.setStandardButtons(QDialogButtonBox.Ok | QDialogButtonBox.Cancel)
        bb.setObjectName("bb")
        bb.accepted.connect(self.accept)
        bb.rejected.connect(self.reject)
        l.addWidget(bb, 3, 0, 1, 2)
        self.toggle_checkbox = tc = QCheckBox(self)
        tc.setObjectName("toggle_checkbox")
        l.addWidget(tc, 2, 0, 1, 2)

def confirm_with_details(parent, title, msg, det_msg='',
        show_copy_button=True):
    d = MessageBox2(MessageBox.INFO, title, msg, det_msg, parent=parent,
                    show_copy_button=show_copy_button)

    return d.exec_() == d.Accepted
